### PR TITLE
[CT-ORG] Create a pipeline for CT-ORG

### PR DESCRIPTION
**Pull request by Piotr1219**
Issue: https://github.com/TheLion-ai/UMIE_datasets/issues/4

- add CT_ORG pipeline 
- modify reading .nii files to not fail on corrupted files
- modify recolor_masks to be resistant to source and target colors from the same values range

### CODE REVIEW

File: src/constants.py [hunk 1]
```diff
@@ -3,4 +3,4 @@
 IMG_FOLDER_NAME = "Images"  # name of the folder with images in each target dataset
 MASK_FOLDER_NAME = "Masks"  # name of the folder with masks in each target dataset
 
-TARGET_PATH = "./data/"  # name of the folder, where processed outputs will be placed
+TARGET_PATH = "/Users/piotr/Desktop/UMIE/processed_data"  # name of the folder, where processed outputs will be placed
```
Comment (andrekomor) [inline src/constants.py:6]: :)

File: config/runner_config.py [hunk 2]
```diff
@@ -119,4 +120,11 @@
             target_path=TARGET_PATH,
         ),
     ),
+    CTORGPipeline(
+        path_args=PathArgs(
+            source_path="",
+            masks_path="",
+            target_path=TARGET_PATH,
```
Comment (andrekomor) [inline config/runner_config.py:127]: Same as in other pipelines
```suggestion
            target_path=TARGET_PATH,
            masks_path="",
```

File: src/pipelines/ct_org.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    """Extractor for study IDs specific to the Knee Osteoarthritis dataset."""
+
+    def _extract(self, img_path: str) -> str:
+        """Extract study id from img path."""
+        study_id = os.path.basename(img_path).replace("-", "_").split("_")[-2]
+        return study_id
+
+
+@dataclass
+class CTORGPipeline(BasePipeline):
```
Comment (andrekomor) [inline src/pipelines/ct_org.py:48]: It is good practice to use abbreviations as other words for readability reasons (Only first letter large) [(Source)](https://www.approxion.com/capital-offenses-how-to-handle-abbreviations-in-camelcase/). Please change everywhere this is used.
image

```suggestion
class CtOrgPipeline(BasePipeline):
```

File: testing/tests/test_check_ct_org.py
Comment (andrekomor) [inline testing/tests/test_check_ct_org.py:1]: rename file to `test_ct_org.py`

File: testing/tests/test_check_ct_org.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
+"""
+test_coronahack.
```
Comment (andrekomor) [inline testing/tests/test_check_ct_org.py:2]: Update

File: testing/tests/test_check_ct_org.py [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+from base.pipeline import PathArgs
+from src.pipelines.ct_org import CTORGPipeline
+from testing.libs.dataset_testing_lib import DatasetTestingLibrary
+
+source_path = os.path.join(os.getcwd(), "testing/test_dummy_data/14_ct_org/input")
+target_path = os.path.join(os.getcwd(), "testing/test_dummy_data/14_ct_org/output")
+masks_path = os.path.join(os.getcwd(), "testing/test_dummy_data/14_ct_org/input")
+expected_output_path = os.path.join(os.getcwd(), "testing/test_dummy_data/14_ct_org/expected_output")
+
```
Comment (andrekomor) [inline testing/tests/test_check_ct_org.py:21]: Add initial clean_up step (It is implemented already in other test files)

File: testing/tests/test_check_ct_org.py [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+def test_ct_org_verify_images_correct():
+    """Test to verify whether all images have contents as expected."""
+    expected_file_tree = glob.glob(f"{str(expected_output_path)}/**/*.png", recursive=True)
+    current_file_tree = glob.glob(f"{str(target_path)}/**/*.png", recursive=True)
+
+    if not DatasetTestingLibrary.verify_all_images_identical(expected_file_tree, current_file_tree):
+        pytest.fail("CT-ORG pipeline created image contents different than expected.")
+
```
Comment (andrekomor) [inline testing/tests/test_check_ct_org.py:55]: Add check jsonl step

**Review by andrekomor — CHANGES_REQUESTED** [review]
Make sure to update README file also

**Review by andrekomor — APPROVED** [review]

File: testing/tests/test_ct_org.py [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+def test_ct_org_verify_images_correct():
+    """Test to verify whether all images have contents as expected."""
+    expected_file_tree = glob.glob(f"{str(expected_output_path)}/**/*.png", recursive=True)
+    current_file_tree = glob.glob(f"{str(target_path)}/**/*.png", recursive=True)
+
+    if not DatasetTestingLibrary.verify_all_images_identical(expected_file_tree, current_file_tree):
+        pytest.fail("CT-ORG pipeline created image contents different than expected.")
+
+
+def test_chest_xray14_verify_jsonl_correct():
```
Comment (andrekomor) [inline testing/tests/test_ct_org.py:63]: Incorrect name

File: testing/libs/dataset_testing_lib.py [hunk 8]
```diff
@@ -69,13 +69,15 @@ def clean_up(directory_to_erase: Path) -> None:
         """Recursively remove files from directory_to_erase."""
         files_to_erase = glob.glob(f"{str(directory_to_erase)}/**", recursive=True)
         dirs_to_erase = []
+        print(files_to_erase)
```
Comment (andrekomor) [inline testing/libs/dataset_testing_lib.py:72]: Remove

File: src/pipelines/ct_org.py [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+        ("create_blank_masks", CreateBlankMasks),
+        # ("delete_imgs_with_no_annotations", DeleteImgsWithNoAnnotations),
+        ("delete_temp_png", DeleteTempPng),
+    )
+    dataset_args: DatasetArgs = field(default_factory=lambda: dataset_config.ct_org)
+    pipeline_args: PipelineArgs = field(
+        default_factory=lambda: PipelineArgs(
+            zfill=2,
+            # Image id is in the source file name after the last underscore
+            img_id_extractor=ImgIdExtractor(),  #
```
Comment (andrekomor) [inline src/pipelines/ct_org.py:69]: Remove `#`

**Review by andrekomor — CHANGES_REQUESTED** [review]

**Review by A-Huli — APPROVED** [review]

**Pull request merged**